sql/core/src/main/scala/org/apache/spark/sql/execution/window/WindowFunctionFrame.scala

```diff
@@ -196,24 +216,15 @@ class FrameLessOffsetWindowFunctionFrame(
     offset: Int,
     ignoreNulls: Boolean = false)
   extends OffsetWindowFunctionFrameBase(
-    target, ordinal, expressions, inputSchema, newMutableProjection, offset) {
+    target, ordinal, expressions, inputSchema, newMutableProjection, offset, ignoreNulls) {
 
-  override def prepare(rows: ExternalAppendOnlyUnsafeRowArray): Unit = {
-    resetStates(rows)
-    if (ignoreNulls) {
-      if (Math.abs(offset) > rows.length) {
-        fillDefaultValue(EmptyRow)
-      } else {
-        findNextRowWithNonNullInput()
-      }
-    } else {
-      // drain the first few rows if offset is larger than zero
-      while (inputIndex < offset) {
```

Comment:
I'll review super carefully this time. So before the refactor, we don't call `fillDefaultValue` if `ignoreNulls` is false. Can you explain why your refactor is safe to change it?

Reply:
It's true. Before the refactor, we don't call `fillDefaultValue` if `ignoreNulls` is false.
Then we increase the `inputIndex`.
If offset is negative, so the `inputIndex` is negative too. We will call `fillDefaultValue` at the write phase.
If offset is positive and greater than or equals to `rows.length`, `inputIndex` is the same as offset value. We will `fillDefaultValue` at the write phase too.
Otherwise, `inputIndex` is in range [0, offset). We could iterator inputIterator safely.